File: 백준/Silver/11663. 선분 위의 점/선분 위의 점.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>


static int lowerBound(int, const std::vector<int>&);
static int upperBound(int, const std::vector<int>&);

int main()
{
	std::ios::sync_with_stdio(false);
	std::cin.tie(nullptr);

	int N, M;
	std::cin >> N >> M;
	std::vector<int> dots(N);

	for (int i = 0; i < N; i++)
	{
		std::cin >> dots[i];
	}

	std::sort(dots.begin(), dots.end());
	std::string output;

	for (int i = 0; i < M; i++)
	{
		int start, end;
		std::cin >> start >> end;

		output += std::to_string((upperBound(end, dots) - lowerBound(start, dots) + 1)) + "\n";
	}

	std::cout << output;
}

static int lowerBound(int start, const std::vector<int>& dots)
{
	int min = 0;
	int max = dots.size();

	while (min < max)
	{
		int mid = (min + max) / 2;

		if (dots[mid] < start)
		{
			min = mid + 1;
		}
		else {
			max = mid;
		}
	}

	return min;
}

static int upperBound(int end, const std::vector<int>& dots)
{
	int min = 0;
	int max = dots.size();

	while (min < max)
	{
		int mid = (min + max) / 2;

		if (dots[mid] <= end)
		{
			min = mid + 1;
		}
		else {
			max = mid;
		}
	}

	return min - 1;
}

File: 백준/Silver/11663. 선분 위의 점/선분 위의 점.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.StringTokenizer;

public class Main {

    static int[] dots;
    static int N, M;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();


        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        dots = new int[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            dots[i] = Integer.parseInt(st.nextToken());
        }

        // 오름차순 정렬
        Arrays.sort(dots);

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            int startIdx = findIdx(start, 0);
            int endIdx = findIdx(end, 1);

            sb.append(endIdx - startIdx + 1).append("\n");
        }

        System.out.println(sb);
    }

    public static int findIdx(int pos, int type) {
        int min = 0;
        int max = dots.length - 1;
        int result = -1;

        if (type == 0) {
            while (min <= max) {
                int mid = (min + max) / 2;

                if (dots[mid] > pos) {
                    max = mid - 1;
                    result = mid;
                }
                else if (dots[mid] < pos) {
                    min = mid + 1;
                    result = min;
                }
                else {
                    return mid;
                }
            }
        } else {
            while (min <= max) {
                int mid = (min + max) / 2;

                if (dots[mid] > pos) {
                    max = mid - 1;
                    result = max;
                }
                else if (dots[mid] < pos) {
                    min = mid + 1;
                    result = mid;
                }
                else {
                    return mid;
                }
            }
        }

        return result;
    }
}